src/controllers/opportunities.js

```diff
@@ -0,0 +1,248 @@
+
+import { OpportunityDto } from '../dto/opportunity.js';
+
+/**
+ * Opportunities controller.
+ * @param {DataAccess} dataAccess - Data access.
+ * @returns {object} Opportunities controller.
+ * @constructor
+ */
+function OpportunitiesController(dataAccess) {
+  if (!isObject(dataAccess)) {
+    throw new Error('Data access required');
+  }
+  const { Opportunity } = dataAccess;
+  if (!isObject(Opportunity)) {
+    throw new Error('Data access required');
+  }
+
+  /**
+   * Gets all opportunities for a given site.
+   * @param {Object} context of the request
+   * @returns {Promise<Response>} Array of opportunities response.
+   */
+  const getAllForSite = async (context) => {
+    const siteId = context.params?.siteId;
+
+    if (!hasText(siteId)) {
+      return badRequest('Site ID required');
+    }
+
+    const opptys = (await Opportunity.allBySiteId(siteId))
+      .map((oppty) => OpportunityDto.toJSON(oppty));
+
+    return ok(opptys);
+  };
+
+  /**
+   * Gets all opportunities for a given site type filtering by status.
+   * @param {Object} context of the request
+   * @returns {Promise<Response>} Array of opportunities response.
+   */
+  const getByStatus = async (context) => {
+    const siteId = context.params?.siteId;
+    const status = context.params?.status || undefined;
+
+    if (!hasText(siteId)) {
+      return badRequest('Site ID required');
+    }
+    if (!hasText(status)) {
+      return badRequest('Status required');
+    }
+
+    const opptys = (await Opportunity.allBySiteIdAndStatus(siteId, status))
+      .map((oppty) => OpportunityDto.toJSON(oppty));
+
+    return ok(opptys);
+  };
+
+  /**
+   * Gets an opportunity for a given site type and opportunity ID.
+   * @param {Object} context of the request
+   * @returns {Promise<Response>} Opportunity response.
+   */
+  const getByID = async (context) => {
+    const siteId = context.params?.siteId;
+    const opptyId = context.params?.opportunityId || undefined;
+
+    if (!hasText(siteId)) {
+      return badRequest('Site ID required');
+    }
+
+    if (!hasText(opptyId)) {
+      return badRequest('Opportunity ID required');
+    }
+
+    const oppty = await Opportunity.findById(opptyId);
+    if (!oppty || oppty.getSiteId() !== siteId) {
+      return notFound('Opportunity not found');
+    }
+    return ok(OpportunityDto.toJSON(oppty));
+  };
+
+  /**
+   * Creates an opportunity
+   * @param {Object} context of the request
+   * @return {Promise<Response>} Opportunity response.
+   */
+  const createOpportunity = async (context) => {
+    const siteId = context.params?.siteId;
+    if (!hasText(siteId)) {
+      return badRequest('Site ID required');
+    }
+    if (!isNonEmptyObject(context.data)) {
+      return badRequest('No data provided');
+    }
+
+    context.data.siteId = siteId;
+    try {
+      const oppty = await Opportunity.create(context.data);
+      return createResponse(OpportunityDto.toJSON(oppty), 201);
+    } catch (e) {
```

Comment:
bad request would only be applicable if a validation error occurs, but there might also be other errors

Reply:
DONE. 
We need to decide what to do with data access errors. Do we return them as they come, or do we transform the message somehow?